test: allow all functional tests to be run or skipped with --usecli

mzumsande opened the pull request:
Fixes #32264

I looked into all current failures listed in the issue, as well all tests that are already disabled for the cli with `self.supports_cli = False`. There are several reasons why existing tests fail  with `--usecli` on many systems, the most important ones are:

- Most common reason is that the test executes a RPC call with a large arg that exceeds `MAX_ARG_STRLEN` of the OS, which is usually 128kb on linux: This is fixed by using `-stdin` for these large calls (idea by 0xB10C)
- they test specifically the rpc interface - nothing to do there except disabling.
- Some functional test submit wrong types to params on purpose to test the error message (which is different when using the cli) - deactivated these specific subtests locally for the cli when there is just one or two of them, deactivated the entire tests when there are more spots
- When python sets `None` for an arg, the cli converts this to 'null' in `arg_to_cli`. This is fine e.g. for boolean args, but doesn't work for strings where it's interpreted as the string 'null'. Bypass this for named args by not including args in case the value is `None` for the cli is used (it's effectively the same as leaving the optional arg out).
-  the `height_or_hash` param used in some RPC needs to be converted to a JSON (effectively adding full quotes).
- Some tests were marked with `self.supports_cli = False` in the past but run fine on master today - enabled those.

In total, this PR fixes all tests that fail on master and reduces the number of tests that are deactivated (`self.supports_cli = False`) from 40 to 21.
It also adds `--usecli` to one CI job (multiprocess, i686, DEBUG) to detect regressions.

mzumsande committed:
test: rename .rpc to ._rpc and remove unnecessary uses

It is usually not necessary, and makes it impossible to use --usecli

mzumsande committed:
test: skip subtests that check for wrong types with cli

The error messages returned for wrong type differ betwen
using rpc or cli - skip these subtests where they happen.

mzumsande committed:
test: Disable several (sub)tests with cli

Reason for each test:
rpc_whitelist.py: Relies on direct RPC calls
wallet_encryption.py: Null characters cannot be passed to suprocess.Popen
wallet_fundrawtransaction.py: multiple checks for wrong types, which have different error messages with cli
wallet_send.py: multiple checks for wrong types

mzumsande committed:
test: add function to convert to json for height_or_hash params

This is necessary for bitcoin-cli because a string without quotes
is not a valid json.

fanquake commented:
If this ends up fixing all issues, should add `--usecli` to atleast one CI, so we catch regressions?

mzumsande commented on test/functional/test_framework/util.py:
```diff
@@ -603,3 +603,8 @@ def sync_txindex(test_framework, node):
     test_framework.wait_until(lambda: node.getindexinfo("txindex")["txindex"]["synced"])
     test_framework.log.debug(f"Synced in {time.time() - sync_start} seconds")
+
+def add_quotes_for_cli(cli, text):
```
Hmm, this is only an issue because  `hash_or_height` expects a JSON object, not a string (and a text without quotes is no valid JSON object), so regular string args don't need this. I think the current naming is confusing, maybe renaming the function "`string_to_json_for_cli`" and calling `json.dumps(text)` inside instead of adding quotes (in case the cli is used) would be better to convey why this is done? Or is there a better solution?

mzumsande replied:
renamed to `convert_to_json_for_cli`

mzumsande committed:
test: convert tuple to json for cli

This makes it possible to run rpc_createmultisig.py with --usecli.

mzumsande committed:
test: make rpc_psbt.py usable with --usecli

The psbt string would include a "=" sign, which would
make the cli interpret this as a named argument.
Fix this by making it an actual named arg with the
correct name.

mzumsande commented:
> If this ends up fixing all issues, should add --usecli to atleast one CI, so we catch regressions?

Yes - it'd be a bit of  a waste of a CI job if we end up disabling too many of the tests with `--usecli`, but depending on how many can be fixed (I'm currently still going through all the existing disabled tests) that would make sense to me.

mzumsande committed:
test: Don't send empty named args with cli

If python passed None for an optional (i.e. 'null' is
sent), this will lead to the arg being interpreted as not
provided by bitcoind - except for string args, for which the arg is
interpreted as as 'null' string. Bypass this by not sending
named args to bitcoin-cli - so that the default value will
actually be used.

Also drops an unnecessary str() conversion, kwargs keys
are always strings.

mzumsande committed:
test: use -stdin for large rpc commands

Because of the MAX_ARG_STRLEN limit (128kb on most systems)
for args, these would usually fail. As a workaround, use
-stdin for these large calls. Idea by 0xB10C.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

mzumsande committed:
test: enable functional tests with large rpc args for cli

Also, the following tests (for which self.supports_cli = False was not
set) will now work with --usecli:

feature_fastprune.py
feature_fee_estimation.py
feature_reindex_readonly.py
feature_taproot.py
mempool_package_rbf.py
p2p_net_deadlock.py
p2p_tx_download.py
rpc_packages.py

mzumsande committed:
test: Enable various tests for usage with cli

These tests run successfully on current master without any changes.

mzumsande committed:
ci: use --usecli in one of the CI jobs

This helps avoid regressions (changes to functional tests
that are not compatible with -usecli).

mzumsande commented:
> If this ends up fixing all issues, should add `--usecli` to atleast one CI, so we catch regressions?

I have added `--usecli` to one CI job now (multiprocess, i686, DEBUG)

mzumsande commented:
Also added a commmit to fix the too-large args (https://github.com/bitcoin/bitcoin/pull/32399#issuecomment-2846837427).

pinheadmz commented on test/functional/test_framework/util.py:
```diff
@@ -610,3 +610,8 @@ def sync_txindex(test_framework, node):
     test_framework.wait_until(lambda: node.getindexinfo("txindex")["txindex"]["synced"])
     test_framework.log.debug(f"Synced in {time.time() - sync_start} seconds")
+
+def convert_to_json_for_cli(cli, text):
+    if cli:
+        return json.dumps(text)
+    return text
```
959dc33283be3c465476f604263f7a2de4007da0

Why not add this method to `BitcoinTestFramework` then you won't need to pass `self.options.usecli` every call?

pinheadmz replied:
Is this only ever needed for passing a hash as a hash-or-height argument?

mzumsande replied:
This is only correct for `hash_or_height`, which because of its dual use is defined as a numeric, non-string parameter and is therefore converted from json. It would be incorrect for actual string args:
You can see it directly using `bitcoin-cli` in the console:
You can do
`bitcoin-cli -regtest getblockstats '"3520a1e79bd3df078c762693b5c113469e267176aefca0d3c1dac29b8db6dd33"'`
but
`bitcoin-cli -regtest getblockstats "3520a1e79bd3df078c762693b5c113469e267176aefca0d3c1dac29b8db6dd33"` fails with `Error parsing JSON`.

For actual string args, it is the other way round:
`bitcoin-cli -regtest getblock "3520a1e79bd3df078c762693b5c113469e267176aefca0d3c1dac29b8db6dd33"` succeeds while
`bitcoin-cli -regtest getblock '"3520a1e79bd3df078c762693b5c113469e267176aefca0d3c1dac29b8db6dd33"'` fails.

Since `BitcoinTestFramework` doesn't know that this is a special `hash_or_height` arg, I've done the conversion not there.

pinheadmz reviewed (approved):
ACK 7c13240d7b9a03b77ad84460e80007063367e05a

Reviewed all commits and tested on macos/arm64. Compared `test_runner --usecli` on master to branch. Master failed 17 tests and skipped 58. This branch skipped 36 and failed none. Some of the skipped tests are due to my platform.

This PR touches test code only but also by adding CI coverage requires all future changes to the API to be bitcoin-cli compatible (or otherwise pass new tests with cli)


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 7c13240d7b9a03b77ad84460e80007063367e05a
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgjgv4ACgkQ5+KYS2KJ
yTrLzxAAlAVG0ALW7dHqa0NXSst4Rf2xMTAViSgPeIXVma69PHCGHk0y4J+ee9D1
fQkS2NNLAh57jYxXzHXLVyb4ic3m4mIgIkFwQvnZJ1tj5U7wwju+12Qy24CzOEHc
m0kXE8IbKui5TTbgVR1I+5+uIEE01Nd6/pyJnRMvJpcHYP3uQXVlPdq2oDPK4a1d
JjsA8yfQfLh37Rm31wfX2Q0tDtxwXMmnpsDTHg/B+cwCAES4DWiFmpXrp5G9bW22
IyLHIIAIDkhUP5d9wEx+ydahXPp6T9pbK84+2oNVk2EK+iQrdeM4iP88TCD8p4FP
+wJJIOElXf6njtKIa8hAuq8e8uYmPsT7VwcK8GGrII2eYlInzLyFE6oPjVwdOlbc
IaeurBZIlPho0Gb+iFOgM62p6ctj3eul+nDbS+eKkQ/FvaYpM+jQVjadvfVzNzp3
Gyvx/JwyVJYp8uuwifWPMyFBMsvUqhZ0CFZrKUUT5Vt7apYkjWXxZUyWaBoDdYDH
57o2VHwygtaJ1YygUGM+NH5jX4IZd2EPdCc7wWgVhN/hJRmc61mNQkESv1XDmlE4
46FMvc378djlOZwyXouTJSaVG3iY0tXS51pNulPUIsIOTLUViDCNa3gkApAqc6zO
lyokClFmwkfDDrSS/Zqfv8m+XE4Pu4eS0KE1IQZQ6LqpFGg9WIk=
=nd5k
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

pinheadmz reviewed (approved):
ACK 2bf3f907ff54efd4fa22930a81269d568e29274e

Trivial changes since last ack: rebase on master, adjust for removed legacy wallet

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 2bf3f907ff54efd4fa22930a81269d568e29274e
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgt3HwACgkQ5+KYS2KJ
yTpQkg/+MxpEt7N0HhnGVSIVABeCyUJjGsKAih1pbNjUlm/aqGdNMj8NWVu4xPeA
IpbzeT9NIMYCdprRcAOlotpGUl5vRpjnWGL5AACqHfP9i5qutjSFgdusYvu5inam
r4TEF2fxI26hjjrVL2VuOyHxMUEUCgI2GLtwVbl7hKK01efKqAw4cCf0KjxYNf/Y
Ja6J8E3QbjeXgRjAY1nkln/ZW+UUGmdArtB8ysZqRO34ch8MznJExJsL13gMzQoZ
QHloensLimqS8XDyVL/FmNGN/GKdQ/sHkUH8BPUUqSdAvRonhb5jA7yNdgzZRArW
2D6Qs0ba2ou5tSUDwC4HyqtaDVumKSGW4wx9P4aDUzY8baXiy7ypoYRzGmB9yT4o
8KW60Ptx2vnjAtMerqWSlR7decyMLQR0CYqh9e3wssUfNrXhhZbh8C3oG0LxhWKZ
nHyora+lTRPLGXvp+q/i4VuvATt74Z/aHdihBhT6P+VOxRhvRZwNuhfjrkrklEs5
BupeG1R3m3sG/ekpdY/Tk4zPRpbFlKksZEiFgvWS0hOHlDq0EQTGWXO83IVwO3sj
RYEOKHevoeKeon7aLJd0hqtOq3Q0sWgHKUm1SmvMSK++HSM54IBfIYc6M1Qn6Fx5
q8XEMSgzixDVF1yhxeD8JGzExVYpX9BJp85NHZ1gVbc0IcwOpwU=
=y2Tc
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

fanquake commented:
>  the CI failure

The `bitcoin` call seems to be missing the binary argument:
```bash
TestFramework (ERROR): Called Process failed with 'Error: Unknown option: -nonamed
...
subprocess.CalledProcessError: Command '['bitcoin', '-m', '-nonamed', '-datadir=/ci_container_base/ci/scratch/test_runner/test_runner_₿_🏃_20250610_161648/feature_fee_estimation_266/node0', '-named', '-stdin']' returned non-zero exit status 1.
```
```bash
./build/bin/bitcoin -m -nonamed        
Error: Unknown option: -nonamed
Try './build/bin/bitcoin --help' for more information.
```

cc @ryanofsky

ryanofsky commented:
> Converting to draft until I understand the CI failure (which I couldn't reproduce locally so far).

Not sure about the error yet but I could reproduce locally by running `BITCOIN_CMD=bitcoin build/test/functional/feature_fee_estimation.py --usecli`

mzumsande commented:
Thanks - will try this fix soon. This has nothing to do with multiprocess by the way - it's my bug, I just happened to pick that CI job to enable the `-usecli` run.

mzumsande commented:
I took @ryanofsky's fix - I had no idea what was going on (#31375 got merged and conflicted with the previous changes), so I guess it's good that I randomly chose to enable `--usecli` in the `multiprocess` CI job.

maflcko reviewed (commented):
review ACK 60106f32852524498d5d6e958a3bcabdfb06ee1a 🕍

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 60106f32852524498d5d6e958a3bcabdfb06ee1a 🕍
q9TcJblhIp5uAW9SpZJvKitcJo0CClUV8fGnmsGoBotY+vf6/F036IIVR9RFmv4HxhGGbteaWo7q+MLE4QZICA==
```

</details>

mzumsande commented on test/functional/rpc_psbt.py:
```diff
@@ -1209,7 +1208,8 @@ def test_psbt_input_keys(psbt_input, keys):
 
-        self.test_sighash_mismatch()
+        if not self.options.usecli:
+            self.test_sighash_mismatch()
```
That's what I did in an earlier version, but with the last push there was a conflict with the recently added `test_sighash_mismatch` subtest (unrelated to https://github.com/bitcoin/bitcoin/commit/62b82fcb410000272340cbdae877ccbbe5da1935) so I disabled only that subtest.
I wonder if there is a better solution for mismatched type checks instead of just disabling them for the cli - but that would probably be something for a follow-up.

mzumsande commented on test/functional/rpc_deprecated.py:
```diff
@@ -32,7 +32,7 @@ def run_test(self):
             self.log.info("Test settxfee RPC deprecation")
             self.nodes[0].createwallet("settxfeerpc")
-            assert_raises_rpc_error(-32, 'settxfee is deprecated and will be fully removed in v31.0.', self.nodes[0].rpc.settxfee, 0.01)
+            assert_raises_rpc_error(-32, 'settxfee is deprecated and will be fully removed in v31.0.', self.nodes[0].settxfee, 0.01)
```
Makes sense - I renamed it to `_rpc`, added a comment and fixed the additional spots. There is a legit for `rpc` use in setting `ensure_ascii`.

mzumsande commented on test/functional/feature_assumeutxo.py:
```diff
@@ -481,7 +482,7 @@ def check_dump_output(output):
         # Use a hash instead of a height
         prev_snap_hash = n0.getblockhash(prev_snap_height)
-        dump_output5 = n0.dumptxoutset('utxos5.dat', rollback=prev_snap_hash)
+        dump_output5 = n0.dumptxoutset('utxos5.dat', rollback=convert_to_json_for_cli(self.options.usecli, prev_snap_hash))
```
Done.

mzumsande commented on test/functional/test_framework/test_node.py:
```diff
@@ -918,7 +918,7 @@ def batch(self, requests):
         """Run bitcoin-cli command. Deserializes returned string as python object."""
         pos_args = [arg_to_cli(arg) for arg in args]
-        named_args = [str(key) + "=" + arg_to_cli(value) for (key, value) in kwargs.items()]
+        named_args = [str(key) + "=" + arg_to_cli(value) for (key, value) in kwargs.items() if value is not None]
```
yes - I removed this in commit "test: allow all functional tests to be run or skipped with --usecli" which touches this line anyway.

mzumsande commented:
[60106f3](https://github.com/bitcoin/bitcoin/commit/60106f32852524498d5d6e958a3bcabdfb06ee1a) to [666016e](https://github.com/bitcoin/bitcoin/commit/666016e56b28b77f798dc85c767b95c1ca0abfae):
Addressed feedback by @maflcko and rebased.

pinheadmz reviewed (approved):
re-ACK 666016e56b28b77f798dc85c767b95c1ca0abfae

Built and ran all functional tests with --usecli on macos/arm64. Reviewed all code changes again since it's been a while since last review. Coverage is really great.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 666016e56b28b77f798dc85c767b95c1ca0abfae
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmhlbmsACgkQ5+KYS2KJ
yTp9HxAAmFHGtNyp2QfIz2EERL9n8BI/sooD3u/8gs6m56qb5Y42qztjq9Qfjk9k
V4fRpNqSpHEByw1UwxLiZyWMz019zI4PzpNXEHjoXalcgzjyKWmm7U9iivBuGOJF
R0tLmtzPxYlNtX8TBGPqcKpOD28VxIvriptHrG5fB3qTQU8NyP0cFdfzVuLj+ASu
JSRUoi7MZWD/yZ+gaF4jDrayVovt5yPMdca6QrVFGL4+pHlEKlwod+WKOjIH+QpN
rcw1FYchw5PHjsppWUKpGymAF44Au/eytzVx9IVf3u7dRrpjyjF9wtJm0yuNnRfm
a2YlV248qjHBa4dQJUuZPEUAiDAXuqe7smNhQAsSkWMVgWYVNlMlsYQNuGlfu4vs
7+waNsqjCc6jkMgNRAhB5WwPgu6KQvaEL2RCKCZxiKNdG+TMPkq789Nv6wbv5hhe
I3Ubg6QrnWCfsamj/DehLcliK15Imt1TPOwZo84ftLOPGAcMlqtmUoz/EOm4235V
k1VtnU3/HQZCkENj1b8YF2LKmRf/XzpwrHXMLS4pwNXV0rbWDD/TA+BtgX73UtZp
Ict2MSNQgjun0wP9w/NxjomOrEfs2R6sNHePXKSlkGh8dyqujoukjmWi435fM8tu
4G9Q7roRbLeaigjH8OnEUBUfjA/tq6/qla+OkloXXWJIpc5AGIQ=
=FOfY
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>
